File: unii-app/lib/data/models/team.dart
class Team {
  const Team({
    required this.id,
    required this.name,
    required this.inviteCode,
    required this.ownerId,
    required this.memberLimit,
    required this.memberCount,
    this.avatarUrl,
    this.description,
    this.myRole,
  });

  factory Team.fromJson(Map<String, dynamic> json) => Team(
    id: (json['id'] as num).toInt(),
    name: json['name'] as String,
    inviteCode: json['invite_code'] as String,
    ownerId: (json['owner_id'] as num).toInt(),
    memberLimit: (json['member_limit'] as num).toInt(),
    memberCount: (json['member_count'] as num).toInt(),
    avatarUrl: json['avatar_url'] as String?,
    description: json['description'] as String?,
    myRole: (json['my_role'] as num?)?.toInt(),
  );

  final int id;
  final String name;
  final String? avatarUrl;
  final String? description;
  final String inviteCode;
  final int ownerId;
  final int memberLimit;
  final int memberCount;

  /// 0 = member, 1 = owner, null = not a member.
  final int? myRole;

  bool get isOwner => myRole == 1;
}

class TeamMember {
  const TeamMember({
    required this.userId,
    required this.role,
    required this.username,
    this.nickname,
    this.avatarUrl,
    this.joinedAt,
  });

  factory TeamMember.fromJson(Map<String, dynamic> json) => TeamMember(
    userId: (json['user_id'] as num).toInt(),
    role: (json['role'] as num).toInt(),
    username: json['username'] as String,
    nickname: json['nickname'] as String?,
    avatarUrl: json['avatar_url'] as String?,
    joinedAt: json['joined_at'] as String?,
  );

  final int userId;
  final int role;
  final String username;
  final String? nickname;
  final String? avatarUrl;
  final String? joinedAt;

  bool get isOwner => role == 1;
}

File: unii-server/src/dto/team.rs
use serde::{Deserialize, Serialize};

use crate::model::team::{TeamMemberWithUser, TeamRow};

#[derive(Debug, Deserialize)]
pub struct CreateTeamReq {
    pub name: String,
    pub description: Option<String>,
    pub avatar_url: Option<String>,
    pub member_limit: Option<i32>,
}

#[derive(Debug, Deserialize)]
pub struct JoinTeamReq {
    pub invite_code: String,
}

#[derive(Debug, Deserialize)]
pub struct TransferOwnerReq {
    pub new_owner_id: i64,
}

#[derive(Debug, Serialize)]
pub struct TeamView {
    pub id: i64,
    pub name: String,
    pub avatar_url: Option<String>,
    pub description: Option<String>,
    pub invite_code: String,
    pub owner_id: i64,
    pub member_limit: i32,
    pub member_count: i64,
    pub my_role: Option<i16>,
}

impl TeamView {
    pub fn build(row: &TeamRow, member_count: i64, my_role: Option<i16>) -> Self {
        Self {
            id: row.id,
            name: row.name.clone(),
            avatar_url: row.avatar_url.clone(),
            description: row.description.clone(),
            invite_code: row.invite_code.clone(),
            owner_id: row.owner_id,
            member_limit: row.member_limit,
            member_count,
            my_role,
        }
    }
}

#[derive(Debug, Serialize)]
pub struct MemberView {
    pub user_id: i64,
    pub role: i16,
    pub username: String,
    pub nickname: Option<String>,
    pub avatar_url: Option<String>,
    pub joined_at: Option<chrono::DateTime<chrono::Utc>>,
}

impl From<TeamMemberWithUser> for MemberView {
    fn from(m: TeamMemberWithUser) -> Self {
        Self {
            user_id: m.user_id,
            role: m.role,
            username: m.username,
            nickname: m.nickname,
            avatar_url: m.avatar_url,
            joined_at: m.joined_at,
        }
    }
}

pub fn is_valid_visibility(v: &str) -> bool {
    matches!(v, "public" | "private")
}

pub fn validate_name(s: &str) -> bool {
    let t = s.trim();
    !t.is_empty() && t.chars().count() <= 50
}
